keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry/README.md

```diff
@@ -0,0 +1,593 @@
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+Alice is setting up a development environment on Kubernetes.
+
+Alice creates a local Kind cluster. She follows the Kind-specific instructions
+for setting up a local registry during cluster creation.
+
+Alice will have some tool she interacts with that builds images and pushes them
+to a registry the cluster can access. That tool might be an IDE like VSCode,
+an infra-as-code toolchain like Pulumi, or a multi-service dev environment like Tilt.
+
+On startup, the tool should connect to the cluster and read the `local-registry`
+ConfigMap. The ConfigMap embeds a versioned config.
+
+If the config specifies a registry location, the tool may automatically adjust
+image pushes to push to the specified registry.
+
+If the config specifies a `help` URL, the tool may prompt the user to set up a
+registry for faster development.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Users may see the ConfigMap and draw mistaken conclusions about how to
+  interact with it. They think deleting the ConfigMap would delete the local
+  registry (which it does not).
+
+- This KEP does not specify any automatic mechanism for keeping the
+  LocalRegistryStatus up-to-date with how the cluster is configured.
+  For example, the user might delete the registry. The cluster might
+  not have a way of knowing that the registry has died.
+  
+- This KEP creates the new API version `local-registry.x-k8s.io/v1alpha1`,
+  This may open a can of worms. On the one hand, cluster implementations want
+  confidence that they're implementing a shared standard. On the other hand,
+  a new API version implies standards of governance, ownership, and API machinery
+  that might not make sense yet.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+None
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Tools that configure a local registry should apply the following ConfigMap to
+the cluster:
+
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: local-image-loaders
```

Comment:
do you have long term examples of what else can be hosted here?
would it only contain "status" type API?

Reply:
yes! I asked the Kind and Minikube folks about this.
https://minikube.sigs.k8s.io/docs/handbook/pushing/
https://kind.sigs.k8s.io/docs/user/quick-start/#loading-an-image-into-your-cluster

In particular, I could imagine a future proposal like

```
loadCommand.v1:
  cmd: "kind load kind load docker-image $IMAGE_REF"
```

or

```
localDockerEnv.v1:
  DOCKER_HOST: ...
```

I don't have a strong sense of whether this belongs in the KEP or not. I could convince myself either way. https://github.com/kubernetes-sigs/kind/issues/1543#issuecomment-628101414

I wrote a brief note on this in the Alternatives section. Is there a place where a more fleshed-out discussion of this should go?

Reply:
these are API choices that have to be made early on.

"local registry status" being under "local-image-loaders" might confuse some consumers with respect to naming.
i preferred the old scheme where the config map was named "local-cluster-registry" (IIRC) and "status.v1" was the only data structure.

ideally there shouldn't be too many separate structures under the config map. i'd consider 2 distinct ones at maximum,  where the second might required because we named the initial structure poorly. structure makes sense only if it would contain multiple fields. it's important to note that this is all golang source code that users would have to care about, validate, default and pick a version for.

for example, "status" can be renamed to "registry.v1" and include the above docker env, load commands and the status fields (potentially under a sub-struct "status" too).
why isn't this desired vs having separate single field structures?
what prevents us from adding the load command and docker env fields today?

i think some more time needs to be spent on the naming and planning what the structures will include.

Reply:
re: "these are API choices that have to be made early on."

Totally agree!

I'm increasingly convinced this is a good API.

That said, if there are existing Kubernetes conventions that I don't know about, it may be worth accepting a sub-optimal API to conform better to what Kubernetes users expect.

Speaking to why it's the right API:

As a tool developer, I want to get my users' images into the cluster as fast as possible. This is a constraint satisfaction problem. I will always want to read all of the options to solve it. e.g.,
- A docker-env makes sense if the user doesn't have another image builder
- A kind load call makes sense if you're loading the image for the first time
- A local registry push makes sense if you're iterating on an image that you've pushed before

These options are orthogonal and (in many cases) interchangeable. If a cluster doesn't support a local registry, it might support the other two options. If we need to version one (e.g., change localRegistryStatus.v1 -> localRegistryStatus.v2), it wouldn't affect the other two options

The reason that I don't want to codify load commands here isn't because I think this is the wrong API. It's because I suspect cluster maintainers are understandably nervous about committing to "drop-in interchangeability" on these kinds of load commands between clusters, c.f.
https://twitter.com/BenTheElder/status/1262460086611394560

Local registries are the one thing that seem to have critical mass right now and we can guarantee interchangeability for.

All of that said, I would accept a "local-cluster-registry" ConfigMap if "configmaps should only have 1 or 2 separate structures underneath them" was an existing convention that's worth following here. (I don't need to see a written guideline or anything; if you say that's what you see working best, I believe you)

Reply:
> All of that said, I would accept a "local-cluster-registry" ConfigMap if "configmaps should only have 1 or 2 separate structures underneath them" was an existing convention that's worth following here. (I don't need to see a written guideline or anything; if you say that's what you see working best, I believe you)

there is no written convention for this as we are inventing the "structureName.vX in a CM" format on the fly. in fact i'm sure a number of people in core k8s will disagree with it, but that's a compromise for not hosting the API and machinery in a repository.

i think the cap of 2 structures in the config map is something that we definitely need to consider...and that is mainly because of:
>> it's important to note that this is all golang source code that users would have to care about, validate, default and pick a version for.

the API we are releasing is a MAJOR release, so we have to be careful what is included from the get-go, as we don't want to increment this document often with changes for single fields. API releases should include substantial mass, so changes should be done in bulks for vNext and multiple existing consumers should participate in future review.

i see some recent discussion in the kind thread for potentially adding another "host" field. maybe this implies to set a deadline of e.g. 1-2 more weeks before solidifying the specification here?

if you think we should keep `local-image-loaders` (as CM name) and `localRegistryStatus` (struct name) for the first release and potentially add another structure for commands / local env (e.g. localEnvironment.v1) in the future, i think this is fine. but if there is more to be added ideally it should go under existing structures, to keep downstream consumers relatively happy.

pelase let me know if you agree with my concerns here.

Reply:
OK, I'm convinced. Ya, I think your concerns are valid. I'm going to make it a single ConfigMap. 

I can think of other ways to address the image-loader decision-tree problem. I marked this out of scope, and added some details about it in the Alternatives section.

As a tangental note, the late-breaking addition makes me wonder if we should have a v1beta1 proposal.

I think what we're stuck on is a circular dependency - the cluster maintainers want _some_ consensus  before they implement anything, and the spec authors/reviewers want more confidence before we settle on a v1 api. This seems like a good fit for what v1beta1 is for. But it might be more overhead than we need. What do you think?

The graduation criteria would be:
- start with `localRegistryHosting.v1beta`
- let all the cluster maintainers implement it, wait a couple months
- graduate it to v1 with all cluster maintainers consenting
wdyt?

Reply:
i'm going to reply in the main thread.